fish-rust/src/env/environment.rs

```diff
@@ -371,8 +379,315 @@
+    // If that has the same UID that we use, we assume the data is correct.
+    if let Some(user_var) = user_var {
+        let unam_narrow = wcs2zstring(&user_var.as_string());
+        let retval = unsafe {
+            libc::getpwnam_r(
+                unam_narrow.as_ptr(),
+                userinfo.as_mut_ptr(),
+                buf.as_mut_ptr(),
+                buf.len(),
+                &mut result,
+            )
+        };
+        if retval == 0 && !result.is_null() {
+            let userinfo = unsafe { userinfo.assume_init() };
+            if unsafe { *result }.pw_uid == uid {
+                if vars.get_unless_empty(L!("HOME")).is_none() {
+                    if !userinfo.pw_dir.is_null() {
+                        let s = unsafe { CStr::from_ptr(userinfo.pw_dir) };
+                        vars.set_one(
+                            L!("HOME"),
+                            EnvMode::GLOBAL | EnvMode::EXPORT,
+                            str2wcstring(s.to_bytes()),
+                        );
+                    } else {
+                        vars.set_empty(L!("HOME"), EnvMode::GLOBAL | EnvMode::EXPORT);
+                    }
+                }
+                return;
+            }
+        }
+    }
+
+    // Either we didn't have a $USER or it had a different uid.
+    // We need to get the data *again* via the uid.
+    let retval = unsafe {
+        libc::getpwuid_r(
+            uid,
+            userinfo.as_mut_ptr(),
+            buf.as_mut_ptr(),
+            buf.len(),
+            &mut result,
+        )
+    };
+    if retval == 0 && !result.is_null() {
+        let userinfo = unsafe { userinfo.assume_init() };
+        let s = unsafe { CStr::from_ptr(userinfo.pw_name) };
+        let uname = str2wcstring(s.to_bytes());
+        vars.set_one(L!("USER"), EnvMode::GLOBAL | EnvMode::EXPORT, uname);
+        // Only change $HOME if it's empty, so we allow e.g. `HOME=(mktemp -d)`.
+        // This is okay with common `su` and `sudo` because they set $HOME.
+        if vars.get_unless_empty(L!("HOME")).is_none() {
+            if !userinfo.pw_dir.is_null() {
+                let s = unsafe { CStr::from_ptr(userinfo.pw_dir) };
+                vars.set_one(
+                    L!("HOME"),
+                    EnvMode::GLOBAL | EnvMode::EXPORT,
+                    str2wcstring(s.to_bytes()),
+                );
+            } else {
+                // We cannot get $HOME. This triggers warnings for history and config.fish already,
+                // so it isn't necessary to warn here as well.
+                vars.set_empty(L!("HOME"), EnvMode::GLOBAL | EnvMode::EXPORT);
+            }
+        }
+    } else if vars.get_unless_empty(L!("HOME")).is_none() {
+        // If $USER is empty as well (which we tried to set above), we can't get $HOME.
+        vars.set_empty(L!("HOME"), EnvMode::GLOBAL | EnvMode::EXPORT);
+    }
+}
+
+/*
+fn setup_path() {
```

Comment:
probably better to remove commented code

Reply:
> I will introduce that as a new commit that can be squashed later since

In general, I prefer squashing eagerly, to incrementally move towards the ideal patch series instead of a big-bang approach.
Also this has the advantage that it helps any reader who doesn't have a good recollection of the previous state (so new revewers but also often the author and previous readers).
If needed I can still see what changed with `git range-diff` etc.

> confstr is available in libc on macOS https://github.com/rust-lang/libc/pull/2883, but we would need to make a PR to add it for other platforms

let's just preserve existing behavior for now, we can put it into `compat.c`

Reply:
Okay after some investigation, I have found the following:

`fish` threats `PATH=""` as equal to `PATH="."`, see https://github.com/fish-shell/fish-shell/issues/3914 and 

https://github.com/fish-shell/fish-shell/blob/0f19d7118b10ee4354317888963c7291334e0173/fish-rust/src/env/environment.rs#L171

so this is only reachable if you specifically _unset_ `PATH`. That makes the use of `get_unless_empty` misleading, since it is never empty. 

Is seems like this is intended to actually threat the unset and empty string in the same way?

Adding `panic!("the disco");` to within the `if path.is_none()` does not cause the tests to fail, so this is not covered by our tests. 

How to test manually:
```fish
set PATH_TO_FISH_EXE "$PWD/build/fish"
sh -c "env -u PATH $PATH_TO_FISH_EXE"
```

As far as I can tell the `strconf` was introduced in https://github.com/fish-shell/fish-shell/commit/290d07a833d3c55654e38a69934e3e022bd968fa

Reply:
Okay I squashed the fallback into the wrong commit (will fix), but I have now added a fallback using `compat.c`

Reply:
> That makes the use of get_unless_empty misleading, since it is never empty.
>
> Is seems like this is intended to actually treat the unset and empty string in the same way?

yeah, we often want to do that

we can add a test like

```
env -u PATH $fish -c 'echo $PATH'
# CHECK: ???
```

but it's not clear what to check for so maybe leave it

Reply:
I added this:
```
# FIXME: Rationalize behavior when PATH is explicitly unset, should this not behave like PATH=""?
# "" is threated like ".", see https://github.com/fish-shell/fish-shell/issues/3914
env -u PATH $fish --no-config -c 'contains -- /usr/bin $PATH; and contains -- /bin $PATH'
and echo PATH has something when unset
# CHECK: PATH has something when unset
```

Reply:
the default path might technically be platform dependent (unless POSIX specifies this?) so I'm not sure if we can write a meaningful test

Reply:
It should be equal to `getconf PATH`, with that info we can actually test this